File: src/com/suushiemaniac/lang/json/lang/JsonReader.java
package com.suushiemaniac.lang.json.lang;

import com.suushiemaniac.lang.json.JSON;
import com.suushiemaniac.lang.json.lang.antlr.JSONBaseVisitor;
import com.suushiemaniac.lang.json.lang.antlr.JSONLexer;
import com.suushiemaniac.lang.json.lang.antlr.JSONParser;
import com.suushiemaniac.lang.json.util.StringUtils;
import com.suushiemaniac.lang.json.value.*;
import com.suushiemaniac.lang.json.value.JsonArray;
import com.suushiemaniac.lang.json.value.JsonObject;
import org.antlr.v4.runtime.ANTLRInputStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;

import java.util.*;
import java.util.stream.Collectors;

public class JsonReader extends JSONBaseVisitor<JSON> {
    public JSON parse(String json) {
        JSONLexer lexer = new JSONLexer(new ANTLRInputStream(json));
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        JSONParser parser = new JSONParser(tokens);
        ParseTree tree = parser.json();
        return this.visit(tree);
    }

    @Override
    public JSON visitStrType(JSONParser.StrTypeContext ctx) {
        return new JsonString(ctx.STRING().getText());
    }

    @Override
    public JSON visitBoolType(JSONParser.BoolTypeContext ctx) {
        return new JsonBoolean(Boolean.parseBoolean(ctx.getText().toLowerCase()));
    }

    @Override
    public JSON visitNumType(JSONParser.NumTypeContext ctx) {
        String numLiteral = ctx.NUMBER().getText();
        return numLiteral.contains(".") ? new JsonFloat(Float.parseFloat(numLiteral)) : new JsonInteger(Integer.parseInt(numLiteral));
    }

    @Override
    public JSON visitNullType(JSONParser.NullTypeContext ctx) {
        return new JsonNull();
    }

    @Override
    public JSON visitArray(JSONParser.ArrayContext ctx) {
        List<JSON> elements = ctx.value().stream().map(this::visit).collect(Collectors.toList());

        return new JsonArray(elements);
    }

    @Override
    public JSON visitObject(JSONParser.ObjectContext ctx) {
        SortedMap<String, JSON> members = new TreeMap<>();

        for (JSONParser.PairContext prCtx : ctx.pair()) {
            JSON valValue = this.visit(prCtx.value());

            members.put(StringUtils.jsonUnwrap(prCtx.STRING().getText()), valValue);
        }

        return new JsonObject(members);
    }
}

File: src/main/kotlin/com/suushiemaniac/lang/json/lang/JsonReader.kt
package com.suushiemaniac.lang.json.lang

import com.suushiemaniac.lang.json.JSON
import com.suushiemaniac.lang.json.antlr.JSONBaseVisitor
import com.suushiemaniac.lang.json.antlr.JSONLexer
import com.suushiemaniac.lang.json.antlr.JSONParser
import com.suushiemaniac.lang.json.util.StringUtils.jsonUnwrap
import com.suushiemaniac.lang.json.value.*
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream

class JsonReader : JSONBaseVisitor<JSON>() {
    fun parse(json: String): JSON {
        val lexer = JSONLexer(CharStreams.fromString(json))
        val tokens = CommonTokenStream(lexer)
        val parser = JSONParser(tokens)

        val tree = parser.json()

        return this.visit(tree)
    }

    override fun visitStrType(ctx: JSONParser.StrTypeContext): JSON {
        return JsonString(ctx.STRING().text)
    }

    override fun visitBoolType(ctx: JSONParser.BoolTypeContext): JSON {
        return JsonBoolean(java.lang.Boolean.parseBoolean(ctx.text.toLowerCase()))
    }

    override fun visitNumType(ctx: JSONParser.NumTypeContext): JSON {
        val numLiteral = ctx.NUMBER().text
        return if (numLiteral.contains("")) JsonFloat(numLiteral.toFloat()) else JsonInteger(numLiteral.toInt())
    }

    override fun visitNullType(ctx: JSONParser.NullTypeContext): JSON {
        return JsonNull.INST
    }

    override fun visitArray(ctx: JSONParser.ArrayContext): JSON {
        val elements = ctx.value().map { this.visit(it) }
        return JsonArray(elements.toMutableList())
    }

    override fun visitObject(ctx: JSONParser.ObjectContext): JSON {
        val members = ctx.pair().map { it.STRING().text.jsonUnwrap() to this.visit(it.value()) }.toMap()
        return JsonObject(members.toSortedMap())
    }
}